### PR TITLE
[sparse] Handle general batch dimensions in bcoo todense/fromdense

### PR DESCRIPTION
**Pull request by jakevdp**

### CODE REVIEW

**Review by froystig — APPROVED** [review]

File: jax/experimental/sparse/bcoo.py [hunk 1]
```diff
@@ -55,6 +55,18 @@
 from jax._src.util import canonicalize_axis
 
 
+def _bcoo_batch_dims_to_front(batched_args, batch_dims):
+  _, indices = batched_args
+  _, indices_bdim = batch_dims
+  n_batch = indices.ndim - 2 + bool(indices_bdim is None)
+
+  if not all(b is None or 0 <= b < n_batch for b in batch_dims):
+    raise NotImplementedError("batch_dims must be None or satisfy 0 < dim < n_batch. "
```
Comment (soraros) [inline jax/experimental/sparse/bcoo.py:64]: @jakevdp Should this have been `0 <= dim < n_batch`?
